Fix Windows build and add CI test suite

spirit1317 opened the pull request:
Fixes #369 Fixes #355

spirit1317 committed:
windows tests summary

spirit1317 committed:
fix __WIN32__ macro problem

spirit1317 committed:
test_context close file handle

spirit1317 committed:
windows tests fix

spirit1317 committed:
appveyor tests API summary

ajoubertza commented:
@spirit1317 Thanks for the PR.  It looks like this is based on the v9.3.2 tag, rather than the head of the develop branch.  Can you please merge the latest develop branch in to this one?  Or better yet, rebase this branch on the head of develop.  I think that will fix the CI problems on Travis related to pytest.

ajoubertza commented on tests/conftest.py:
```diff
@@ -2,4 +2,51 @@
+
+@pytest.hookimpl()
+def pytest_sessionfinish(session):
+    """ collets all tests to be run and outputs to bat script """
```
Typo: collets -> Collects

ajoubertza commented on tests/conftest.py:
```diff
@@ -2,4 +2,51 @@
+    """ collets all tests to be run and outputs to bat script """
+    if '--collect-only' in sys.argv and '-q' in sys.argv and 'nt' in os.name:
+        print("Generating windows test script...")
+        script_path = os.path.join(os.path.dirname(__file__),'run_tests_win.bat')
```
There are many PEP8 violations in this file.  Please run something like `black tests/conftest.py` to fix it.  Or `flake8` to point them out, and fix manually.  In this line, missing a space after the comma. 

It looks like the automatic Codacy quality check is no longer working for PRs.  It has been "missing" for the last few PRs.

ajoubertza commented on appveyor.yml:
```diff
@@ -132,8 +131,20 @@ build_script:
-after_build:
+test_script:
+  - "%PYTHON_ROOT%/Scripts/pip.exe install gevent pytest psutil six numpy"
+  - cmd: if %PYTHON_VER%==27 "%PYTHON_ROOT%/Scripts/pip.exe" install enum34 futures trollius zipp
```
We shouldn't need to install all these packages manually, since they are in `setup.py` as test requirements. 

It should work if we use:
```
python setup.py test '-c pytest_empty_config.txt --collect-only -q -q tests'
```
instead of
```
pytest.exe -c pytest_empty_config.txt --collect-only -q -q tests
```

The benefit of using `python setup.py test` is that it reduces maintenance.   If we change requirements in  `setup.py` in future, we don't have to change `appveyor.yml` as well. 

The downside of using `python setup.py test`, is that it will install all those packages for every single test.  That will probably increase the time to run the tests unnecessarily.

Also note that there are some specific version limits in [`setup.py`](https://github.com/tango-controls/pytango/blob/cb9b202e6451b744e2a1847c1ae49b213e521fc1/setup.py#L488-L498) - we must match that in the AppVeyor config.  For both Python 2 and Python 3.

Conclusion:  install the test requirements manually.  However, make a change to use the correct package versions.

ajoubertza commented on tests/conftest.py:
```diff
@@ -2,4 +2,51 @@
+                summary.append(test)
+                f.seek(0)
+                f.write(json.dumps(summary))
+                f.truncate()
```
Missing newline at the end of the file.  Please add one.

ajoubertza commented on appveyor.yml:
```diff
@@ -132,8 +131,20 @@ build_script:
+  - cmd: nul > summary.json
+  - cmd: run_tests_win.bat
+  - cmd: set API_URL=%APPVEYOR_API_URL%/api/tests/batch
+  - cmd: powershell -command "$body = Get-Content summary.json -Raw; Invoke-Webrequest $env:API_URL -Method 'POST' -UseBasicParsing -ContentType 'application/json; charset=utf-8' -Body $body"
```
This is nice! 

I see on AppVeyor that 75 to 77 of the tests have failed, yet the build is still green.   Two questions:
- Can we make the overall build fail (red), if any tests fail?
- What is causing those test failures?

ajoubertza commented on appveyor.yml:
```diff
@@ -80,18 +80,18 @@ init:
+  - cmd: cd "C:\projects\"
+  - cmd: md tango_test
+  - cmd: cd "C:\projects\tango_test"
+  - appveyor DownloadFile https://github.com/spirit1317/tango-test-exe/raw/main/TangoTest.exe
```
Not ideal, but we can use this until there is a better solution.

In the readme on that repo, could you add the exact commands you used to build the executable, please?

ajoubertza reviewed (changes requested):
Nice work.  Most import change required is to make the AppVeyor build correctly report failure if any tests fail.

spirit1317 committed:
appveyor tests status

spirit1317 committed:
merge develop

spirit1317 committed:
fixing test issues

spirit1317 commented:
Hello, I changed the code according to your comments:

- codacy improoved
- merged windows_9.3.2 with develop
- added pip packages versions to appveyor.yml
- all tests are passing on Windows, but I had to remove `process=True` from `with DeviceTestContext(process=True)`. Also only one DeviceTestContest can be used per test as you know, so I had to rewrite some tests in test_server.py, but the coverage is the same (please take a look).
- if (in future) some test fail, the appveyor buld will not pass green
- added readme.md to TangoTest.exe repo, but I didn't build TangoTest it myself
- all tests are passing also on Linux on travis, however I had problems with get_attribute_config() on my computer

andygotz commented:
Impressive - well done! I look forward to a new package in pip to try out. How much work is that?

ajoubertza commented on tests/test_server.py:
```diff
@@ -371,7 +371,7 @@ def write(self, attr):
 # Test properties
 
-def test_device_property_no_default(typed_values, server_green_mode):
+def test_device_property(typed_values, server_green_mode):
```
These new versions of the property tests don't quite cover the same functionality as before.  I also prefer the old names for the tests.  Here's an update that uses your ideas, but keeps the test functionality and names the same.

```python
def test_device_property_no_default(typed_values, server_green_mode):
    dtype, values, expected = typed_values
    patched_dtype = dtype if dtype != (bool,) else (int,)
    value = values[1]

    class TestDevice(Device):
        green_mode = server_green_mode

        prop_without_db_value = device_property(dtype=dtype)
        prop_with_db_value = device_property(dtype=dtype)

        @command(dtype_out=bool)
        def is_prop_without_db_value_set_to_none(self):
            return self.prop_without_db_value is None

        @command(dtype_out=patched_dtype)
        def get_prop_with_db_value(self):
            return self.prop_with_db_value

    with DeviceTestContext(TestDevice,
                           properties={'prop_with_db_value': value}) as proxy:
        assert proxy.is_prop_without_db_value_set_to_none()
        assert_close(proxy.get_prop_with_db_value(), expected(value))


def test_device_property_with_default_value(typed_values, server_green_mode):
    dtype, values, expected = typed_values
    patched_dtype = dtype if dtype != (bool,) else (int,)

    default = values[0]
    value = values[1]

    class TestDevice(Device):
        green_mode = server_green_mode

        prop_without_db_value = device_property(
            dtype=dtype, default_value=default
        )
        prop_with_db_value = device_property(
            dtype=dtype, default_value=default
        )

        @command(dtype_out=patched_dtype)
        def get_prop_without_db_value(self):
            return self.prop_without_db_value

        @command(dtype_out=patched_dtype)
        def get_prop_with_db_value(self):
            return self.prop_with_db_value

    with DeviceTestContext(TestDevice,
                           properties={'prop_with_db_value': value}) as proxy:
        assert_close(proxy.get_prop_without_db_value(), expected(default))
        assert_close(proxy.get_prop_with_db_value(), expected(value))
```

ajoubertza commented on tests/test_server.py:
```diff
@@ -560,13 +557,8 @@ class TestDevice(Device):
-                           properties={'prop': value},
-                           process=True) as proxy:
-        assert_close(proxy.get_prop(), expected(value))
-
```
This test case was important.  It checks that mandatory properties that have values specified in the database, work correctly.

Here's is a version that tests both cases:
- works if a value is specified in the database
- raises an error if not specified in the database

We now have to split it into two different tests.

```python
def test_mandatory_device_property_with_db_value_succeeds(
        typed_values, server_green_mode):
    dtype, values, expected = typed_values
    patched_dtype = dtype if dtype != (bool,) else (int,)
    default, value = values[:2]

    class TestDevice(Device):
        green_mode = server_green_mode

        prop = device_property(dtype=dtype, mandatory=True)

        @command(dtype_out=patched_dtype)
        def get_prop(self):
            return self.prop

    with DeviceTestContext(TestDevice,
                           properties={'prop': value}) as proxy:
        assert_close(proxy.get_prop(), expected(value))


def test_mandatory_device_property_without_db_value_fails(
        typed_values, server_green_mode):
    dtype, _, _ = typed_values

    class TestDevice(Device):
        green_mode = server_green_mode
        prop = device_property(dtype=dtype, mandatory=True)

    with pytest.raises(DevFailed) as context:
        with DeviceTestContext(TestDevice):
            pass
    assert 'Device property prop is mandatory' in str(context.value)

```

ajoubertza commented on appveyor.yml:
```diff
@@ -153,8 +153,8 @@ build_script:
 test_script:
-  - "%PYTHON_ROOT%/Scripts/pip.exe install \"gevent!=1.5a1\" \"pytest<5\" psutil \"six>=1.10\" \"numpy>=1.1\""
-  - cmd: if %PYTHON_VER%==27 "%PYTHON_ROOT%/Scripts/pip.exe" install enum34 futures trollius "zipp>=0.5,<2" "pyparsing<3"
+  - "%PYTHON_ROOT%/Scripts/pip.exe install \"gevent!=1.5a1\" pytest psutil \"six>=1.10\" \"numpy>=1.1\" enum34" 
```
Thanks for the fixes to the versions.

Note that `enum34` is only for Python 2.  It is a back-port of the enum code from Python 3.4, so needed on versions of Python < 3.4.

ajoubertza commented on setup.py:
```diff
@@ -414,6 +415,37 @@ def has_html(self):
     sub_commands.append(('install_html', has_html))
 
 
+class CheckTestsErrors(Command):
```
Nice idea to do this check via Python code.  A lot easier than from the Windows command line!

ajoubertza commented on setup.py:
```diff
@@ -414,6 +415,37 @@ def has_html(self):
+    description = "Checks tests summary.json for failed tests and raises errror if found"
+    
+    user_options = [
+        ("summary-file=",None,"Path to summary.json"),
+        ("ignore-gevent-errors",None,"Failed tests using GreenMode.Gevent will not raise error."),
```
Please run a linter over you new changes - should be spaces after commas in the tuples, for PEP8.  Also in the lists and function argument lists below.

ajoubertza commented on setup.py:
```diff
@@ -414,6 +415,37 @@ def has_html(self):
+    def run(self):
+        with open(self.summary_file,"r") as f:
+            summary = f.read()
+            summary = json.loads(summary)
```
This is fine, but note that `json.load` will read and load from the file in one step (instead of a `read` and `loads`).

ajoubertza commented on setup.py:
```diff
@@ -585,7 +617,8 @@ def setup_args():
         'install_html': install_html,
-        'install': install}
+        'install': install,
+        'check_tests_errors': CheckTestsErrors}
```
I see all the other command classes are using lower case.  How about we keep this one lower case as well, to match that convention?  `check_tests_errors`.

spirit1317 committed:
test fixes, test summary check, required packages

spirit1317 commented:
Hi,

I compled with your latest remarks. Additionally:

- On some environments few Gevent tests were failing. You asked to drop for now Gevent support for those enviroments and let the build pass with those errors. I use setup.py command `check_tests_errors` with flag `--ignore-gevent-errors` in combination with [GEVENT_NOT_SUPPORTED](https://github.com/spirit1317/pytango/blob/3502427ef34445cc0d5c9dbc5ef53dc62590a61e/appveyor.yml#L51) `appveyor.yml` environment variable. This variable can be deleted when we find the cause of those errors.

- I tried to debug the issue with failing `test_identity_command` with Gevent. Sometimes one of the `bytes_devstring` from test_utils.TYPED_VALUES in the list of 10 same elements gets truncated and end letters are changed. What I found is the command arguments are not passed to device command correctly. It happens after "allways_executed_hook" is called, but this function does nothing so, its not causing the problem. Following the call stack:
[connection.py:137](https://github.com/spirit1317/pytango/blob/3502427ef34445cc0d5c9dbc5ef53dc62590a61e/tango/connection.py#L137) is ok
[server.py:1142](https://github.com/spirit1317/pytango/blob/3502427ef34445cc0d5c9dbc5ef53dc62590a61e/tango/server.py#L1142) the args strings are truncated
Please see next comment.

spirit1317 commented:
Heres the simple code, that can replicate the issue from previous comment without using pytest:
```python
import sys
import six

from tango import GreenMode
from tango.server import Device
from tango.server import command
from tango.test_utils import DeviceTestContext, TYPED_VALUES

values = TYPED_VALUES[(str,)]

class TestDevice(Device):
    green_mode = GreenMode.Gevent

    @command(dtype_in=(str,), dtype_out=(str,))
    def identity(self, arg):
        return arg

with DeviceTestContext(TestDevice, timeout=600) as proxy:
    for value in values:
        result = proxy.identity(value)
```

Heres a screen shot of full debug stack (list of strings shown in args is already truncated on this line):
![call_stack](https://user-images.githubusercontent.com/41666803/102074637-5a8ed200-3e05-11eb-99a5-fcb49e85f12b.png)

ajoubertza commented:
Thanks for the details on the gevent issue.  We can fix it one day, if someone really needs the support.

ajoubertza reviewed (approved):
Looks good - nice work, @spirit1317!

I changed the target of the PR merge to the `develop` branch, instead of `stable`.  The diff looks much better now.  We only update `stable` when making a release.
